### ISSUE TITLE
language: support optional markers for dynamic fields

### ISSUE THREAD

**Issue by disintegrator** [cue-lang/cue#2107]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version v0.4.3 darwin/arm64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What did you do?

https://cuelang.org/play/?id=QTJz8OOP282#cue@export@cue

```cue
#Geos: ["UK", "AU"]
#PerGeo: {
  _value: _

  for _, geo in #Geos {
    (geo)?: _value
  }
}

out: #PerGeo & { _value: >= 0.0 }
```

### What did you expect to see?

I was expecting to get the same output as the following: https://cuelang.org/play/?id=U2gP5VJNJS-#cue@export@cue

```cue
out: {}
```

or more accurately

### What did you see instead?

The fields that resulted from the comprehension were not marked as optional. The optional marker was ignored and I got the following:

```cue
out: {
	UK: >=0.0
	AU: >=0.0
}
```

---

Keywords: string interpolation, field name expression, optional fields, structs

**Comment by disintegrator** [comment]
A smaller reproducer from @rogpeppe: https://tip.cuelang.org/play/?id=gGtOGipIe0B#cue@export@json

**Comment by disintegrator** [comment]
Another reproducer from @rogpeppe and @mvdan: https://tip.cuelang.org/play/?id=MTJhnnEvkU8#cue@export@cue

**Comment by myitcv** [comment]
Marking as a feature request, because this is simply not supported yet. Arguably there should be a syntax error in the linked example until it is supported.

**Comment by disintegrator** [comment]
> Marking as a feature request, because this is simply not supported yet. Arguably there should be a syntax error in the linked example until it is supported.

Marking it as a feature request makes sense to but I do think it's a bug to allow an unsupported syntax to pass silently.

**Comment by mpvl** [comment]
Marked this as v0.6. With the eminent changes to optional field representations, this will be easy to make work. So rather than throwing an error, we will make this work.

For now, you can write this as:
```
[geo]: _value
```

**Comment by disintegrator** [comment]
Thanks @mpvl! I'll keep on eye on it.

**Comment by myitcv** [comment]
@disintegrator - just to flag this got completed in https://github.com/cue-lang/cue/commit/c3138e362ec88061b8a83368ccc66511581a7263. Our bot doesn't do a good job of notifying when things get done!

**Comment by disintegrator** [comment]
@myitcv amazing, thank you all!

**Comment by jpluscplusm** [comment]
@myitcv thanks to any and all involved in adding this feature :-) May I ask which release you anticipate it'll be available in?

**Comment by myitcv** [comment]
@jpluscplusm it's already present in tip at the commit indicate, which means it's available in the pseudo version v0.6.0-0.dev.0.20230323113716-c3138e362ec8. The next version we will be tagging based on tip is an alpha of v0.6.0.
